src/infra/marketplace/marketplaceCmdRepo.go

```diff
@@ -0,0 +1,256 @@
+			return errors.New("InstallRequiredServiceError: " + err.Error())
+		}
+	}
+
+	return nil
+}
+
+func (repo *MarketplaceCmdRepo) addMissingOptionalDataFieldsToMap(
+	receivedDataFieldsMap *map[string]string,
+	requiredDataFields []valueObject.MarketplaceCatalogItemDataField,
+) {
+	for _, requiredDataField := range requiredDataFields {
+		if requiredDataField.IsRequired {
+			continue
+		}
+
+		requiredKeyStr := requiredDataField.Key.String()
+		if len((*receivedDataFieldsMap)[requiredKeyStr]) != 0 {
+			continue
+		}
+
+		requiredDefaultValueStr := requiredDataField.DefaultValue.String()
+		(*receivedDataFieldsMap)[requiredKeyStr] = requiredDefaultValueStr
+	}
+}
+
+func (repo *MarketplaceCmdRepo) parseCmdStepsWithDataFields(
+	cmdSteps []valueObject.MarketplaceItemInstallStep,
+	dataFieldsMap map[string]string,
+) ([]valueObject.MarketplaceItemInstallStep, error) {
+	cmdStepsWithDataFields := []valueObject.MarketplaceItemInstallStep{}
+
+	for _, cmdStep := range cmdSteps {
+		cmdStepStr := cmdStep.String()
+		cmdStepDataFieldKeys, _ := infraHelper.GetAllRegexGroupMatches(
+			cmdStepStr,
+			`%(.*?)%`,
+		)
+
+		for _, cmdStepDataFieldKey := range cmdStepDataFieldKeys {
+			dataFieldValue := dataFieldsMap[cmdStepDataFieldKey]
+			cmdStepWithDataFieldStr := strings.ReplaceAll(
+				cmdStepStr,
+				"%"+cmdStepDataFieldKey+"%",
+				dataFieldValue,
+			)
+			cmdStepStr = cmdStepWithDataFieldStr
+		}
+
+		cmdStepWithDataField, _ := valueObject.NewMarketplaceItemInstallStep(cmdStepStr)
+		cmdStepsWithDataFields = append(cmdStepsWithDataFields, cmdStepWithDataField)
+	}
+
+	return cmdStepsWithDataFields, nil
+}
+
+func (repo *MarketplaceCmdRepo) runCmdStepsWithDataFields(
+	catalogCmdSteps []valueObject.MarketplaceItemInstallStep,
+	catalogDataFields []valueObject.MarketplaceCatalogItemDataField,
+	receivedDataFields []valueObject.MarketplaceInstallableItemDataField,
+	installDir valueObject.UnixFilePath,
+) error {
+	receivedDataFieldsMap := map[string]string{}
+
+	for _, receivedDataField := range receivedDataFields {
+		receivedDataFieldKeyStr := receivedDataField.Key.String()
+		receivedDataFieldsMap[receivedDataFieldKeyStr] = receivedDataField.Value.String()
+	}
+
+	receivedDataFieldsMap["installDirectory"] = installDir.String()
+	receivedDataFieldsMap["installUuid"] = uuid.New().String()[:16]
+	repo.addMissingOptionalDataFieldsToMap(
+		&receivedDataFieldsMap,
+		catalogDataFields,
+	)
+
+	cmdStepWithDataFields, err := repo.parseCmdStepsWithDataFields(
+		catalogCmdSteps,
+		receivedDataFieldsMap,
+	)
+	if err != nil {
+		return errors.New("ParseCmdStepWithDataFieldsError: " + err.Error())
+	}
+
+	for _, cmdStepWithDataField := range cmdStepWithDataFields {
+		cmdStepWithDataFieldStr := cmdStepWithDataField.String()
+		_, err = infraHelper.RunCmdWithSubShell(cmdStepWithDataFieldStr)
+		if err != nil {
+			return errors.New(
+				"RunCmdStepError (" + cmdStepWithDataFieldStr + "): " + err.Error(),
+			)
+		}
+	}
+
+	return nil
+}
```

Comment:
Vou descrever o que eu entendi do seu código:

- Cria um mapa de data fields recebido;
- Popula o mapa com a chave e valor (isso não deveria ser um GetAsMap() do VO MarketplaceInstallableItemDataField?)
- Injeta o installDirectory e installUuid no mapa de data fields recebido;
- Add campos faltantes opcionais do data fields (?) (nome estranho)

Dentro desse método com nome estranho (addMissingOptionalDataFieldsToMap):
-- Verifica se o requiredDataField é requerido? (?) (se o campo é requiredDataField pq você verifica se IsRequired?)
-- Se o campo requerido possui um valor (?) (O que é esse pedaço de código?)
```
		requiredKeyStr := requiredDataField.Key.String()
		if len((*receivedDataFieldsMap)[requiredKeyStr]) != 0 {
			continue
		}
```
-- Cria o "requiredDefaultValueStr" (desde quando required tem default value? wtf?) e injeta ele no mapa receivedDataFieldsMap?
Conclusão: entendi nada e você saiu injetando valor pra dentro do seu ponteiro. **Caos total.**

Seguindo de volta no método principal:
- Você em teoria "parseia" o cmdSteps com os data fields, o que está implícito uma vez que um dos params do cmd steps é data fields;

Dentro desse método parseCmdStepsWithDataFields:
-- Você faz um for para todos os cmdSteps;
-- Dentro do for você busca os placeholders dos data fields que no caso é a chave;
-- Para cada chave você consome um dataFieldsMap em busca da chave, só que não verifica se é que a chave existe (erro de lógica);
-- Substitui a chave pelo valor dela;
-- Recria o VO InstallStep (conflito de nome CmdStep com InstallStep);
-- Popula o cmdStepsWithDataFields.
Conclusão: Eu entendi o que você fez pois eu entendo o contexto, sem o contexto eu teria tido dificuldade.
1. O seu objetivo não é "parse", parse é converter de um formato pra outro. O seu objetivo é substituir placeholders. Logo, o nome do seu método deveria ser "replaceCmdStepsPlaceholders(cmdSteps, receivedDataFields)" ou "interpolateCmdSteps(cmdSteps, receivedDataFields)".
2. Anteriormente você criou o map[string]string de dataFields com um objetivo de injetar o installDirectory e installUuid nele. E daí passa essa lógica temporária pro seu método auxiliar parseCmdSteps. Você começou uma lógica de transformação num método e espalhou essa lógica pra um segundo método. Isso torna a leitura muito difícil.
O que você deveria ter feito é criado o installDirectory e installUuid como VOs do tipo MarketplaceInstallableItemDataField e colocado eles dentro da sua receivedDataFields. Daí em seguida, você usava o GetAsMap() que eu sugeri dentro da sua interpolateCmdSteps. Isso faria UM MUNDO de diferença pra leitura da sua intenção, pois a lógica temporária de transformação estaria contida dentro de um método só.

Seguindo de volta no método principal:
- Você executa o cmdStepWithDataFields (que pode se chamar preparedCmdSteps por exemplo pra indicar que eles estão prontos pra rodar);
- Caso algum deles dê erro, você interrompe a execução colando o comando inteiro na mensagem (não sei se é uma boa ideia, melhor usar o índice pra você saber QUAL etapa parou);

Reply:
Fiz algumas modificações, mas restaram algumas ressalvas:
- Não consigo enxergar como seria feito o GetAsMap() dentro do meu VO, tendo em vista que meu receivedDataFields é um slice de VOs, ou seja, meu GetAsMap() estaria dentro de uma única instância de VO e não do meu slice em si. Entendo que a ideia do GetAsMap() to VO é transformar meu slice de VOs em um map[string]string como eu já faço, mas não vejo como isso funcionaria de um em um.
Se a ideia era, na verdade, o meu GetAsMap() transformar o meu VO unitário, não o slice todo, em um map[string]string não seria um ganho muito alto dado como está sendo feito o parse do slice para um map (um for com 2 linhas).
- O parse para map[string]string não é apenas para injeção do installDir e do installUuid, mas para também facilitar a leitura e usabilidade do receivedDataFields, mas percebi como isso prejudicou a leitura.
- No método addMissingOptionalDataFieldsToMap a ideia é injetar todos os default values dos data fields que não são obrigatórios no manifesto. Ou seja, se o usuário não enviar, utilizaremos os default values. Para facilitar a leitura troquei para "getMissingOptionalDataFields" (pensei em "getMissingCatalogItemOptionalDataFields" mas acho que ficaria muito grande), retornei os data fields que estão faltando no receivedDataFields com os default values e concatenei com o receivedDataFields original.